src/runmode-af-xdp.c

```diff
@@ -0,0 +1,424 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \ingroup afxdppacket
+ *
+ * @{
+ */
+
+/**
+ * \file
+ *
+ * \author Richard McConnell <richard_mcconnell@rapid7.com>
+ *
+ * AF_XDP socket runmode
+ *
+ */
+#define PCAP_DONT_INCLUDE_PCAP_BPF_H 1
+#define SC_PCAP_DONT_INCLUDE_PCAP_H  1
+#include "suricata-common.h"
+#include "tm-threads.h"
+#include "conf.h"
+#include "runmodes.h"
+#include "runmode-af-xdp.h"
+#include "output.h"
+#include "log-httplog.h"
+#include "detect-engine-mpm.h"
+
+#include "alert-fastlog.h"
+#include "alert-debuglog.h"
+
+#include "flow-bypass.h"
+
+#include "util-debug.h"
+#include "util-time.h"
+#include "util-cpu.h"
+#include "util-affinity.h"
+#include "util-device.h"
+#include "util-runmodes.h"
+#include "util-ioctl.h"
+#include "util-ebpf.h"
+#include "util-byte.h"
+
+#include "source-af-xdp.h"
+
+#ifdef HAVE_AF_XDP
+#include <linux/if_xdp.h>
+#include <linux/if_link.h>
+#include <xdp/xsk.h>
+#endif
+
+const char *RunModeAFXDPGetDefaultMode(void)
+{
+    return "workers";
+}
+
+void RunModeIdsAFXDPRegister(void)
+{
+    RunModeRegisterNewRunMode(
+            RUNMODE_AFXDP_DEV, "single", "Single threaded af-xdp mode", RunModeIdsAFXDPSingle);
+    RunModeRegisterNewRunMode(RUNMODE_AFXDP_DEV, "workers",
+            "Workers af-xdp mode, each thread does all"
+            " tasks from acquisition to logging",
+            RunModeIdsAFXDPWorkers);
+    RunModeRegisterNewRunMode(RUNMODE_AFXDP_DEV, "autofp",
+            "Multi socket AF_XDP mode.  Packets from "
+            "each flow are assigned to a single detect "
+            "thread.",
+            RunModeIdsAFXDPAutoFp);
+    return;
+}
+
+#ifdef HAVE_AF_XDP
+
+#define DEFAULT_BUSY_POLL_TIME    20
+#define DEFAULT_BUSY_POLL_BUDGET  64
+#define DEFAULT_GRO_FLUSH_TIMEOUT 2000000
+#define DEFAULT_NAPI_HARD_IRQS    2
+
+static void AFXDPDerefConfig(void *conf)
+{
+    AFXDPIfaceConfig *pfp = (AFXDPIfaceConfig *)conf;
+    /* Pcap config is used only once but cost of this low. */
+    if (SC_ATOMIC_SUB(pfp->ref, 1) <= 1) {
+        SCFree(pfp);
+    }
+}
+
+static int ConfigSetThreads(AFXDPIfaceConfig *aconf, const char *entry_str)
+{
+    SCEnter();
+    const char *active_runmode = RunmodeGetActive();
+
+    if (active_runmode && !strcmp("single", active_runmode)) {
+        aconf->threads = 1;
+        SCReturnInt(0);
+    }
+
+    if (entry_str == NULL) {
+        SCLogError(SC_ERR_INVALID_VALUE, "Number of threads for interface \"%s\" not specified",
+                aconf->iface);
+        SCReturnInt(-EINVAL);
```

Comment:
Looks strange to have errno stype value here.

Reply:
Can be adapted to the more commonly used ``TmEcode`` style without any issue.